File: examples/language-features/VbNetNet10Latest/Vb15_5/NonTrailingNamedArguments/NonTrailingNamedArguments.vb
Option Strict On

Namespace Vb15_5.NonTrailingNamedArguments
    Public Module Formatter
        Public Function Format(text As String,
                               Optional upper As Boolean = False,
                               Optional suffix As String = "") As String
            Dim result As String = If(upper, text.ToUpperInvariant(), text)
            Return result & suffix
        End Function

        ' VB names arguments with := rather than C#'s colon. Before 15.5 a named
        ' argument had to be followed only by other named ones; now one may be
        ' named in its own position with the rest left positional.
        Public Function NamedFirst() As String
            Return Format(text:="value", False, "!")
        End Function

        Public Function NamedInMiddle() As String
            Return Format("value", upper:=False, "!")
        End Function

        ' A fully-named call may be reordered, which VB has always allowed.
        Public Function AllNamedReordered() As String
            Return Format(suffix:="?", text:="value", upper:=True)
        End Function
    End Module
End Namespace

File: examples/language-features/CSharp/dotnet/10/13.0/library/CSharp7_2/NonTrailingNamedArguments/NonTrailingNamedArguments.cs
namespace Net10_CSharp13_Library.CSharp7_2.NonTrailingNamedArguments
{
    public class Formatter
    {
        public static string Format(string text, int width, bool upper)
        {
            string result = upper ? text.ToUpperInvariant() : text;
            return result.PadRight(width, '.');
        }

        // Before C# 7.2 a named argument had to be followed only by other named
        // arguments. Now one may be named in its own position and the rest left
        // positional — useful for labeling a bare literal.
        public static string NamedFirst()
        {
            return Format(text: "value", 10, false);
        }

        public static string NamedInMiddle()
        {
            return Format("value", width: 10, false);
        }

        // The C# 7.2 rule only constrains a named argument that is followed by
        // positional arguments: that name must sit in the position its
        // parameter actually occupies. It says nothing about a call where
        // every argument is named — that has been legal, and freely
        // reorderable, since C# 4.0, because there are no positional
        // arguments left to find a parameter for.
        public static string AllNamed()
        {
            return Format(upper: true, text: "value", width: 10);
        }
    }
}
